### ISSUE TITLE
StackOverflowError()

### ISSUE THREAD

**Issue by jtschneider** [julia-vscode/LanguageServer.jl#957]
Hello all

I encounter a `StackOverflowError()` that causes the julia executable for the Julia VS code extension to escalate memory usage to 20 - 25GB of RAM. Most importantly, the language server simply fails more than 5 times but does not get killed properly.

Here is my error message from VS code's Extension output console for "Julia Language Server":

```julia
[ Info: Starting the Julia Language Server
[ Info: Symbol server store is at '/home/jan/.config/Code/User/globalStorage/julialang.language-julia/symbolstorev5'.
Internal error: encountered unexpected error during compilation of getproperty:
StackOverflowError()
unknown function (ip: 0x7f920a288414)
unknown function (ip: 0x7f920a28e90b)
unknown function (ip: 0x7f920a297cdc)
unknown function (ip: 0x7f920a29ccef)
unknown function (ip: 0x7f920a29d083)
unknown function (ip: 0x7f920a346c9e)
unknown function (ip: 0x7f920a3488c0)
unknown function (ip: 0x7f920a2c0f8f)
jl_apply_generic at /usr/bin/../lib/julia/libjulia-internal.so.1 (unknown line)
Internal error: encountered unexpected error during compilation of getproperty:
StackOverflowError()
Internal error: encountered unexpected error during compilation of getproperty:
StackOverflowError()
Internal error: encountered unexpected error during compilation of getproperty:
StackOverflowError()

signal (6): Aborted
in expression starting at /home/jan/.vscode/extensions/julialang.language-julia-1.2.4/scripts/languageserver/main.jl:19
Internal error: encountered unexpected error during compilation of getproperty:
StackOverflowError()
```
Installed is Julia Lang extension v1.2.4 for VS code
Julia v1.6.1

**Comment by davidanthoff** [comment]
I'm also seeing these stack overflows in crash reporting. Should probably be a priority for us to try to fix this.

**Comment by jtschneider** [comment]
Today, I got a slightly different error message after updating the Julia extension on my remote server to which I connect via the VS Code SSH editing extension. This is the output from VS Code Julia Server:

```log
~/.vscode-server/extensions/julialang.language-julia-1.3.29/scripts/languageserver/main.jl:22
julia: /buildworker/worker/package_linux64/build/src/cgmemmgr.cpp:833: virtual uint8_t* {anonymous}::RTDyldMemoryManagerJL::allocateCodeSection(uintptr_t, unsigned int, unsigned int, llvm::StringRef): Assertion `!code_allocated' failed.
```
Could this give a better hint where the `StackOverflowError()` is comming from? Curiously, after restarting VS Code and reconnecting to the remote server, I only encounter the `StackOverflowError()`at the same line of code;
```log
[ Info: Cloud was unable to cache GLFW_jll in the past, we won't try to retrieve it again.
[ Info: Cloud was unable to cache Wayland_jll in the past, we won't try to retrieve it again.
[ Info: Cloud was unable to cache Qt5Base_jll in the past, we won't try to retrieve it again.
[ Info: Cloud was unable to cache LibVPX_jll in the past, we won't try to retrieve it again.
[ Info: Cloud was unable to cache ZeroMQ_jll in the past, we won't try to retrieve it again.
[ Info: Couldn't retrieve cache file for PaddedViews.
[ Info: Couldn't retrieve cache file for ChainRules.
[ Info: Couldn't retrieve cache file for DiffRules.
[ Info: Couldn't retrieve cache file for LoweredCodeUtils.
[ Info: Couldn't retrieve cache file for FuzzyCompletions.
[ Info: Couldn't retrieve cache file for JuliaInterpreter.
[ Info: Couldn't retrieve cache file for StatsFuns.
[ Info: Couldn't retrieve cache file for LibSSH2_jll.
Internal error: encountered unexpected error during compilation of getproperty:
Internal error: encountered unexpected error during compilation of getproperty:
StackOverflowError()

signal (6): Aborted
in expression starting at ~/.vscode-server/extensions/julialang.language-julia-1.3.29/scripts/languageserver/main.jl:22
Internal error: encountered unexpected error during compilation of getproperty:
StackOverflowError()
```

**Comment by jtschneider** [comment]
> Let's hope so. We might be able to get a new insiders release out today, so you could give that a go.

It's looking gooood!! I have no more StackOverflowErrors() !  :partying_face:

**Comment by jtschneider** [comment]
Hello @pfitzseb ,
sorry to disappoint, but I was trying out the new version of the Julia extension on my personal computer (as opposed to the remote machine I was refering to in the previous post) and there I am still encountering the StackOverflowError(). This is after a fresh install of the Julia extension. I deleted ` /home/jan/.vscode/extensions/julialang.language-julia-1.4.0/` and `'/home/jan/.config/Code/User/globalStorage/julialang.language-julia/` before installing the extension.

This is the Julia LanguageServer Output:

```
signal (6): Aborted
in expression starting at /home/jan/.vscode/extensions/julialang.language-julia-1.4.0/scripts/languageserver/main.jl:22
Internal error: encountered unexpected error during compilation of getproperty:
StackOverflowError()

signal (6): Aborted
in expression starting at /home/jan/.vscode/extensions/julialang.language-julia-1.4.0/scripts/languageserver/main.jl:22

signal (6): Aborted
in expression starting at /home/jan/.vscode/extensions/julialang.language-julia-1.4.0/scripts/languageserver/main.jl:22
<unknown>:0: error: starting new .cfi frame before finishing the previous one
<unknown>:0: error: starting new .cfi frame before finishing the previous one
<unknown>:0: error: starting new .cfi frame before finishing the previous one
<unknown>:0: error: starting new .cfi frame before finishing the previous one
<unknown>:0: error: starting new .cfi frame before finishing the previous one
<unknown>:0: error: starting new .cfi frame before finishing the previous one
<unknown>:0: error: this directive must appear between .cfi_startproc and .cfi_endproc directives
<unknown>:0: error: this directive must appear between .cfi_startproc and .cfi_endproc directives
<unknown>:0: error: this directive must appear between .cfi_startproc and .cfi_endproc directives
<unknown>:0: error: this directive must appear between .cfi_startproc and .cfi_endproc directives
<unknown>:0: error: this directive must appear between .cfi_startproc and .cfi_endproc directives
<unknown>:0: error: this directive must appear between .cfi_startproc and .cfi_endproc directives

signal (6): Aborted
in expression starting at /home/jan/.vscode/extensions/julialang.language-julia-1.4.0/scripts/languageserver/main.jl:22

signal (6): Aborted
in expression starting at /home/jan/.vscode/extensions/julialang.language-julia-1.4.0/scripts/languageserver/main.jl:22

signal (6): Aborted
in expression starting at /home/jan/.vscode/extensions/julialang.language-julia-1.4.0/scripts/languageserver/main.jl:22
julia: /buildworker/worker/package_linux64/build/src/cgmemmgr.cpp:833: virtual uint8_t* {anonymous}::RTDyldMemoryManagerJL::allocateCodeSection(uintptr_t, unsigned int, unsigned int, llvm::StringRef): Assertion `!code_allocated' failed.
```

**Comment by pfitzseb** [comment]
Hm. This error has even less useful info :( It would be kinda useful to find out which file causes this -- any chance you can share the code you were working on when the error happened?

**Comment by jtschneider** [comment]
I'm not sure if this in anyway helpful but I was trying to run the language server myself from the console by means of:

```
❯ pwd
/home/jan/.vscode/extensions/julialang.language-julia-1.4.0/scripts/languageserver
❯
❯
❯ julia --startup-file=no --history-file=no --project=/home/jan/.vscode/extensions/julialang.language-julia-1.4.0/scripts/environments/languageserver main.jl /home/jan/.julia/environments/v1.6 --debug=no /tmp/testdir /home/jan/.config/Code/User/globalStorage/julialang.language-julia download --detached=no
[ Info: Starting the Julia Language Server
┌ Error: Some Julia code in the VS Code extension crashed
└ @ Main ~/.vscode/extensions/julialang.language-julia-1.4.0/scripts/error_handler.jl:15
ERROR: Invalid number of arguments passed to julia language server.
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] top-level scope
   @ ~/.vscode/extensions/julialang.language-julia-1.4.0/scripts/languageserver/main.jl:24

```

**Comment by pfitzseb** [comment]
Yeah, that's not gonna work. 
Try installing `StaticLint#master` and then run `StaticLint.lint_file("your/file")`; if that works then it should repro your error.

**Comment by jtschneider** [comment]
It should be the file I am editing in VS code. I checked with this dummy file and I still get the StackOverflow:

```julia
using Prtintf, Dates

pritnln("Hello world")

@sprintf("Hello world %d", 2)
```

**Comment by jtschneider** [comment]
AHA! I am getting somewhere! I think I can pin it down to an environment file that is causing the problem! I am developing a small package and when I open the folder with VS Code, the environment is naturally the local one:

My local package Project TOML:
```toml
name = "LongRangeITensors"
uuid = "eaa3e942-efcb-420d-a848-ae583f98f630"
authors = ["Jan Schneider <j-t-s@outlook.de> and contributors"]
version = "0.1.12"

[deps]
HDF5 = "f67ccb44-e63f-5c2f-98bd-6dc0ccc4ba2f"
ITensors = "9136182c-28ba-11e9-034c-db9fb085ebd5"
LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
Printf = "de0858da-6303-5e67-8744-51eddeeeb8d7"

[compat]
julia = "1.5"

[extras]
Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"

[targets]
test = ["Test"]
```

However, my standard environment `(@1.6)` does contain the following:
```toml
[deps]
BenchmarkTools = "6e4b80f9-dd63-53aa-95a3-0cdb28fa8baf"
Combinatorics = "861a8166-3701-5b0c-9a16-15d98fcdc6aa"
IJulia = "7073ff75-c697-5162-941a-fcdaad2a7d2a"
ITensors = "9136182c-28ba-11e9-034c-db9fb085ebd5"
JuliaFormatter = "98e50ef6-434e-11e9-1051-2b60c6c9e899"
LongRangeITensors = "eaa3e942-efcb-420d-a848-ae583f98f630"
Octavian = "6fd5a793-0b7e-452c-907f-f8bfe9c57db4"
Revise = "295af30f-e4ad-537b-8983-00126c2a3abe"
TimeEvoMPS = "1bec0e0a-d6ff-11e9-2fd9-7928ca571592"
Unitful = "1986cc42-f94f-5a68-af5c-568840ba703d"
```

**Comment by jtschneider** [comment]
> Yeah, that's not gonna work.
> Try installing `StaticLint#master` and then run `StaticLint.lint_file("your/file")`; if that works then it should repro your error.

```julia
julia> using StaticLint

shell> cat testfile.jl
using Prtintf, Dates

pritnln("Hello world")

@sprintf("Hello world %d", 2)

julia> StaticLint.lint_file("testfile.jl")
┌ Warning: LibSSH2_jll not stored on disc
└ @ SymbolServer ~/.julia/packages/SymbolServer/rrlGM/src/SymbolServer.jl:192
ERROR: type FileServer has no field symbolserver
Stacktrace:
 [1] setproperty!(x::StaticLint.FileServer, f::Symbol, v::Dict{Symbol, SymbolServer.ModuleStore})
   @ Base ./Base.jl:34
 [2] setup_server(env::String, depot::String, cache::String)
   @ StaticLint ~/.julia/packages/StaticLint/0OsZh/src/interface.jl:4
 [3] setup_server(env::String) (repeats 2 times)
   @ StaticLint ~/.julia/packages/StaticLint/0OsZh/src/interface.jl:2
 [4] lint_file(rootpath::String)
   @ StaticLint ~/.julia/packages/StaticLint/0OsZh/src/interface.jl:41
 [5] top-level scope
   @ REPL[7]:1

julia>

**Comment by pfitzseb** [comment]
> It should be the file I am editing in VS code. I checked with this dummy file and I still get the StackOverflow:

No, you can basically assume we parse every file in the workspace (and every file "linked" to by those via `include`, in either direction).

**Comment by jtschneider** [comment]
> Yeah, that's not gonna work.
> Try installing `StaticLint#master` and then run `StaticLint.lint_file("your/file")`; if that works then it should repro your error.

That doesn't seem to work. I assume that the function `lint_file(rootpath, server)` takes two arguments, a string to the file location and a `server`, from [here](https://www.julia-vscode.org/StaticLint.jl/dev/syntax/#StaticLint.lint_file)
I was hoping that I could get a more detailed error message with this..?

**Comment by pfitzseb** [comment]
Try using `SymbolServer#master` as well (and maybe a few more dependencies). Can't guarantee that'll work though ;)

**Comment by jtschneider** [comment]
I have found the error. As always, it was a very stupid thing :unamused: 
The bug was indeed related to a file inside my project folder. In a subfolder for backups a scrambled up, unreadable binary file with file ending `.jl` sneaked itself a way into my git ripo. I do not understand how, but apparently it happened. I suppose the Linter then tried it's best on this unreadable mash of bits and simply crashed.

I guess if there are a couple of lessons to be learned: 
1.) take care of your R/W. I have already experienced when my Linux machine goes in hybernate mode and VS code is open and a file is unsaved, it gets scramble and unreadable.
2.) No idea if this is feasible, but could the Linter first check if a `*.jl` file is readable in standard encoding formats? I guess UTF-8 for most Julia files? I would assume this simply safety check would also be helpful for more than my idiotic self.

Best
Jan
